system-addon/lib/MessageCenterRouter.jsm

```diff
@@ -0,0 +1,204 @@
+        params: {url: "https://mozilla.org"}
+      }
+    }
+  },
+  {
+    id: "ONBOARDING_2",
+    template: "simple_snippet",
+    content: {
+      title: "Make Firefox your go-to-browser",
+      body: "It doesn't take much to get the most from Firefox. Just set Firefox as your default browser and put control, customization, and protection on autopilot."
+    }
+  },
+  {
+    id: "ONBOARDING_3",
+    template: "simple_snippet",
+    content: {
+      title: "Did you know?",
+      body: "All human beings are born free and equal in dignity and rights. They are endowed with reason and conscience and should act towards one another in a spirit of brotherhood."
+    }
+  }
+];
+
+/**
+ * getRandomItemFromArray
+ *
+ * @param {Array} arr An array of items
+ * @returns one of the items in the array
+ */
+function getRandomItemFromArray(arr) {
+  const index = Math.floor(Math.random() * arr.length);
+  return arr[index];
+}
+
+/**
+ * @class _MessageCenterRouter - Keeps track of all messages, UI surfaces, and
+ * handles blocking, rotation, etc. Inspecting MessageCenter.state will
+ * tell you what the current displayed message is in all UI surfaces.
+ *
+ * Note: This is written as a constructor rather than just a plain object
+ * so that it can be more easily unit tested.
+ */
+class _MessageCenterRouter {
+  constructor(initialState = {}) {
+    this.initialized = false;
+    this.messageChannel = null;
+    this._state = Object.assign({
+      currentId: null,
+      blockList: {},
+      messages: {}
+    }, initialState);
+    this.onMessage = this.onMessage.bind(this);
+  }
+
+  get state() {
+    return this._state;
+  }
+
+  set state(value) {
+    throw new Error("Do not modify this.state directy. Instead, call this.setState(newState)");
+  }
+
+  setState(callbackOrObj) {
+    const newState = (typeof callbackOrObj === "function") ? callbackOrObj(this.state) : callbackOrObj;
+    this._state = Object.assign({}, this.state, newState);
+    return new Promise(resolve => {
+      this._onStateChanged(this.state);
+      resolve();
+    });
+  }
+
+  _onStateChanged(state) {
+    this.messageChannel.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "ADMIN_SET_STATE", data: state});
+  }
+
+  async sendNextMessage(target, id) {
+    let message;
+    await this.setState(state => {
+      message = getRandomItemFromArray(state.messages.filter(item => item.id !== state.currentId && !state.blockList[item.id]));
+      return {currentId: message ? message.id : null};
+    });
+    if (message) {
+      target.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "SET_MESSAGE", data: message});
+    } else {
+      target.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "CLEAR_MESSAGE"});
+    }
+  }
+
+  async clearMessage(target, id) {
+    if (this.state.currentId === id) {
+      await this.setState({currentId: null});
+    }
+    target.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "CLEAR_MESSAGE"});
+  }
+
+  async onMessage({data: action, target}) {
```

Comment:
Why does this have to be `async` compared to our other `feed.onAction` handlers that may trigger async handlers but the `onAction` method itself is not async?

Reply:
it just makes it a bit easier to test, `onAction` often returns promises (which ends up doing the same thing) but a lot of the feeds were written before we had `async/await` I think

Reply:
👍